src/gc.c

```diff
@@ -3036,6 +3213,118 @@ JL_DLLEXPORT jl_value_t *jl_gc_alloc_3w(void)
+            goto valid_object;
+        }
+        jl_gc_pool_t *pool =
+            jl_all_tls_states[meta->thread_n]->heap.norm_pools +
+            meta->pool_n;
+        if (meta->fl_begin_offset == (uint16_t) -1) {
+            // case 2: this is a page on the newpages list
+            jl_taggedvalue_t *newpages = pool->newpages;
+            // Check if the page is being allocated from via newpages
+            if (!newpages)
+                return NULL;
+            char *data = gc_page_data(newpages);
+            if (data != meta->data) {
+                // Pages on newpages form a linked list where only the
+                // first one is allocated from (see reset_page()).
+                // All other pages are empty.
+                return NULL;
+            }
+            // This is the first page on the newpages list, where objects
+            // are allocated from.
+            if ((char *)cell >= (char *)newpages) // past allocation pointer
+                return NULL;
+            goto valid_object;
+        }
+        // case 3: this is a page with a freelist
+        // marked or old objects can't be on the freelist
+        if (cell->bits.gc)
+            goto valid_object;
+        // When allocating from a freelist, three subcases are possible:
+        // * The freelist of a page has been exhausted; this was handled
+        //   under case 1, as nfree == 0.
+        // * The freelist of the page has not been used, and the age bits
+        //   reflect whether a cell is on the freelist or an object.
+        // * The freelist is currently being allocated from. In this case,
+        //   pool->freelist will point to the current page; any cell with
+        //   a lower address will be an allocated object, and for cells
+        //   with the same or a higher address, the corresponding age
+        //   bit will reflect whether it's on the freelist.
+        // Age bits are set in sweep_page() and are 0 for freelist
+        // entries and 1 for live objects. The above subcases arise
+        // because allocating a cell will not update the age bit, so we
+        // need extra logic for pages that have been allocated from.
+        unsigned obj_id = (off - off2) / osize;
+        // We now distinguish between the second and third subcase.
+        // Freelist entries are consumed in ascending order. Anything
+        // before the freelist pointer was either live during the last
+        // sweep or has been allocated since.
+        if (gc_page_data(cell) == gc_page_data(pool->freelist)
+            && (char *)cell < (char *)pool->freelist)
+            goto valid_object;
+        // We know now that the age bit reflects liveness status during
+        // the last sweep and that the cell has not been reused since.
+        if (!(meta->ages[obj_id / 8] & (1 << (obj_id % 8)))) {
+            return NULL;
+        }
+        // Not a freelist entry, therefore a valid object.
+        valid_object:
```

Comment:
Less indent for the label, see other places in the code.

Reply:
Okay, that's easy enough.